pos_sale_order_delivery_carrier/models/pos_config.py

```diff
@@ -0,0 +1,23 @@
+from odoo import api, fields, models
+
+
+class PosConfig(models.Model):
+    _inherit = "pos.config"
+
+    @api.model
+    def _domain_delivery_carrier_ids(self):
+        return (
+            "[('available_in_pos', '=', True),"
+            "'|',"
+            "('company_id', '=', False),('company_id', '=', company_id)]"
+        )
+
+    allow_delivery_carrier = fields.Boolean(default=True)
+    delivery_carrier_ids = fields.Many2many(
+        comodel_name="delivery.carrier",
+        relation="delivery_carrier_pos_config_rel",
+        column1="pos_config_id",
+        column2="delivery_carrier_id",
+        string="Shipping Methods",
+        domain=lambda self: self._domain_delivery_carrier_ids(),
```

Comment:
If you follow my suggestion about removing `available_in_pos` for `delivery.carrier`, then you can simply change this domain for a `check_company=True` attribute, that's gonna do a better job at ensuring the company matches